url/url_test.cpp

```diff
@@ -36,11 +40,98 @@ int main() {
+            etest::expect(!url->fragment.has_value());
+        }
+
+        url = p.parse("http://bücher.de");
+
+        etest::expect(url.has_value());
+
+        if (url.has_value()) {
+            etest::expect(url->scheme == "http");
+            etest::expect(std::get<std::string>(url->host->data) == "xn--bcher-kva.de");
+            etest::expect(!url->port.has_value());
+            etest::expect(std::get<1>(url->path)[0] == "");
+            etest::expect(!url->query.has_value());
+            etest::expect(!url->fragment.has_value());
+        }
+
+        url = p.parse("https://√.com/i/itunes.gif");
+
+        etest::expect(url.has_value());
+
+        if (url.has_value()) {
+            etest::expect(url->scheme == "https");
+            etest::expect(std::get<std::string>(url->host->data) == "xn--19g.com");
+            etest::expect(!url->port.has_value());
+            etest::expect(std::get<1>(url->path)[0] == "i");
+            etest::expect(std::get<1>(url->path)[1] == "itunes.gif");
+            etest::expect(!url->query.has_value());
+            etest::expect(!url->fragment.has_value());
+        }
+
+        url = p.parse("https://ar.wikipedia.org/wiki/نجيب_محفوظ");
+
+        etest::expect(url.has_value());
+
+        if (url.has_value()) {
+            etest::expect(url->scheme == "https");
+            etest::expect(std::get<std::string>(url->host->data) == "ar.wikipedia.org");
+            etest::expect(!url->port.has_value());
+            etest::expect(std::get<1>(url->path)[0] == "wiki");
+            etest::expect(std::get<1>(url->path)[1] == "%D9%86%D8%AC%D9%8A%D8%A8_%D9%85%D8%AD%D9%81%D9%88%D8%B8");
+            etest::expect(!url->query.has_value());
+            etest::expect(!url->fragment.has_value());
+        }
+
+        url = p.parse("tel:+1-555-555-5555");
+
+        etest::expect(url.has_value());
+
+        if (url.has_value()) {
+            etest::expect(url->scheme == "tel");
+            etest::expect(!url->host.has_value());
+            etest::expect(!url->port.has_value());
+            etest::expect(std::get<0>(url->path) == "+1-555-555-5555");
+            etest::expect(!url->query.has_value());
+            etest::expect(!url->fragment.has_value());
+        }
+
+        url = p.parse("https://www.youtube.com/watch?v=2g5xkLqIElU&list=PLHwvDXmNUa92NlFPooY1P5tfDo4T85ORz&index=3");
+
+        etest::expect(url.has_value());
+
+        if (url.has_value()) {
+            etest::expect(url->scheme == "https");
+            etest::expect(std::get<std::string>(url->host->data) == "www.youtube.com");
+            etest::expect(!url->port.has_value());
+            etest::expect(std::get<1>(url->path)[0] == "watch");
+            etest::expect(url->query == "v=2g5xkLqIElU&list=PLHwvDXmNUa92NlFPooY1P5tfDo4T85ORz&index=3");
+            etest::expect(!url->fragment.has_value());
+        }
+    });
+
+    bool ret = etest::run_all_tests();
+
+    u_cleanup();
```

Comment:
It's required for making asan shut up. i.e only required for tests.

Reply:
Memory leaks? Wouldn't it be required everywhere we call the UrlParser from then, at least before the application is shut down? At least the browser tui is 100% asan and valgrind clean.

Reply:
Required everywhere that ICU gets used, technically. But yeah, I guess that does mean this is an issue for anything that calls UrlParser. I could wrap the cleanup function the same way I wrapped the initialization stuff, but just like with init, there's the question of where to put it and when to call it.

Reply:
See https://unicode-org.github.io/icu/userguide/icu/design.html#icu4c-initialization-and-termination for detail.

```
When an application is terminating it should call the function u_cleanup(), which frees all heap storage and other system resources that are held internally by the ICU library. While the use of u_cleanup() is not strictly required, failure to call it will cause memory leak checking tools to report problems for resources being held by ICU library.
```

Reply:
Wrapping it and placing it in url.h will make it more discoverable and we won't have to add direct dependencies on icu in all tests and applications that use the url parser.

Reply:
I added an `icu_cleanup()` routine to the "url" namespace that needs to be called before you exit tests that use the UrlParser.